File: cSharp/day104.cs
                                                                                            using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections.Generic;

const string DbFile = "__db.csv";

static int AddToCsv(string[] rowValues)
{
    // Replace inner quotes and wrap each value in quotes
    for (int i = 0; i < rowValues.Length; i++)
    {
        rowValues[i] = "\"" + rowValues[i].Replace("\"", "\\\"") + "\"";
    }
    // Join row values into CSV format
    string row = string.Join(",", rowValues) + "\n";

    // Read entire file and count number of newline characters
    string contents = File.ReadAllText(DbFile);

    if (contents.Length > 0) {
        char contentLastCharacter = contents[^1];

        if (contentLastCharacter != '\n') {
            row = '\n' + row;
        }
    }

    // Append row to file (create if not exists)
        File.AppendAllText(DbFile, row, Encoding.UTF8);

    return contents.Count(c => c == '\n');
}

static string[] TrimQuotes(string[] values)
{
    return values.Select(v => v.Trim('"')).ToArray();
}

static string[] ReadCsvRow(int rowId)
{
    string[] rows = File.ReadAllLines(DbFile);
    if (rowId < 0 || rowId >= rows.Length)
        throw new IndexOutOfRangeException("Row ID out of range");

    return TrimQuotes(rows[rowId].Split(","));
}

static int GetCsvRowCount()
{
    return File.ReadAllLines(DbFile).Length;
}

static List<string[]> ReadAllCsvRow()
{
    int numRows = GetCsvRowCount();;

    var rows = new List<string[]>(numRows);
    for (int x = 0; x < numRows; x++)
    {
        rows.Add(ReadCsvRow(x));
    }
    return rows;
}

static int GetCsvColumnNumber(string columnName)
{
    string[] columns = ReadCsvRow(0);
    for (int i = 0; i < columns.Length; i++)
    {
        if (columns[i].Trim() == columnName)
        {
            return i + 1; // 1-based index like Go version
        }
    }
    throw new Exception("Column not found");
}

static string GetCSVCellValue(int rowId, int columnNumber) {
    string[] row = ReadCsvRow(rowId);


    return row[columnNumber - 1];
}

static int[] SearchCsvRows(string columnName, string value) {
    List<string[]> rows = ReadAllCsvRow();

    int columnNumber = GetCsvColumnNumber(columnName);
    int columnIndex = columnNumber - 1;

    Stack<int> result = new Stack<int>();

    for (int x = 1; x < rows.Count; x++)
    {
        string[] row = rows[x];
        if (row[columnIndex] == value)
        {
            result.Push(x);
        }
    }
    return result.ToArray();
}

static int? FindFirstCsvRow(string columnName, string value)
{
    List<string[]> rows = ReadAllCsvRow();

    int columnNumber = GetCsvColumnNumber(columnName);
    int columnIndex = columnNumber - 1;

    Stack<int> result = new Stack<int>();

    for (int rowId = 1; rowId < rows.Count; rowId++)
    {
        string[] row = rows[rowId];
        if (row[columnIndex] == value)
        {
            return rowId;
        }
    }
    return null;
}

void SetCsvToBlank()
{
    string dbFile = "__db.csv";

    try
    {
        // Open or create the file, overwrite existing content (truncate)
        using (var fileStream = new FileStream(dbFile, FileMode.Create, FileAccess.Write))
        {
            // Nothing to write — this effectively clears the file
        }
    }
    catch (Exception ex)
    {
        throw new Exception("Failed to clear CSV file", ex);
    }
}


void DeleteCsvRow(int deleteRowId)
{
    var rows = ReadAllCsvRow();
    SetCsvToBlank();

    for (int rowId = 0; rowId < rows.Count; rowId++)
    {
        var row = rows[rowId];
        if (rowId == deleteRowId)
        {
            continue;
        }
        AddToCsv(row);
    }
}

void SearchRoutine()
{
    Console.WriteLine("Please input your query:");

    string query1 = Console.ReadLine();

    int[] rowIds1 = SearchCsvRows("FirstName", query1);
    int[] rowIds2 = SearchCsvRows("LastName", query1);
    int[] rowIds3 = SearchCsvRows("Email", query1);


    //
    int[] rowIds = rowIds1.Concat(rowIds2).Concat(rowIds3).ToArray();

    Console.WriteLine(rowIds.Length);

    foreach (
        int rowId
        in rowIds
    )
    {
        string[] row = ReadCsvRow(rowId);
        Console.WriteLine(
            String.Join(
                ',',
                row
                )
        );
    }
}

void UpdateCsvRow(
    int rowId,
    string columnName,
    string value)
{
    var rows = ReadAllCsvRow();
    int columnNumber = GetCsvColumnNumber(columnName);
rows[rowId][columnNumber - 1] = value;

    SetCsvToBlank();

    foreach (var row in rows)
    {
        AddToCsv(row);
    }
}




void AddRowRoutine()
{
    Console.WriteLine("Input email:");

    string email1 = Console.ReadLine();

    int? rowId = FindFirstCsvRow("Email", email1);

    Console.WriteLine("Input first name:");
    string firstName = Console.ReadLine();

    Console.WriteLine("Input last name:");
    string lastName = Console.ReadLine();

    if (rowId.HasValue)
    {
        UpdateCsvRow(rowId.Value, "firstName", firstName);
        UpdateCsvRow(rowId.Value, "lastName", lastName);
        return;
    }

    AddToCsv(new[] { email1, firstName, lastName });

    Console.WriteLine("Email/Name successfully added to CSV");
}

string[][] rows = ReadAllCsvRow().ToArray();

if (rows.Length == 0)
{
    return;
}

int[] rowMaxLengths = new int[rows[0].Length];

foreach (string[] row in rows)
{
    foreach (
        var (index, cell)
            in row.Select(
                (arrayValue, Arrayindex)
                    => (Arrayindex, arrayValue)
            ))
    {
        rowMaxLengths[index] = Math.Max(
            rowMaxLengths[index],
            cell.Length
        );
    }
}

int totalLength = rowMaxLengths.Sum();

Console.WriteLine(
    ' ' +
    new string('-', totalLength + 3 + rowMaxLengths.Length)
);

foreach (
    var (rowId, row)
        in rows.Select(
            (arrayValue, Arrayindex)
                => (Arrayindex, arrayValue)
        )
)
{

    Console.Write("| ");
    foreach (
        var (index, cell)
            in row.Select(
                (arrayValue, Arrayindex)
                    => (Arrayindex, arrayValue)
            )
    )
    {
        int padding = rowMaxLengths[index] - cell.Length;
        Console.Write(cell + new string(' ', padding) + " |");
    }
    Console.Write("\n");

    if (rowId == 0) {
        Console.WriteLine(
            new string(
                '-',
                totalLength
                + 5
                + rowMaxLengths.Length
            )
        );
    }
}




Console.WriteLine(
    ' ' +
    new string('-', totalLength + 3 + rowMaxLengths.Length)
);

while (true)
{

    Console.WriteLine("Press 's' to search; 'a' to add; ESC to exit");
    ConsoleKeyInfo keyInfo = Console.ReadKey(intercept: true);

    if (keyInfo.KeyChar == 's')
    {
        SearchRoutine();
        continue;
    }

    
    if (keyInfo.KeyChar == 'a')
    {
        AddRowRoutine();
        continue;
    }



    if (keyInfo.Key == ConsoleKey.Escape)
    {
        Console.WriteLine("\nEscape pressed. Exiting...");
        break;
    }

    Console.WriteLine("Invalid command");
}

File: go/day104.go
package main

import (
	"fmt"
	"os"
	"strings"
	"bytes"
	"bufio"
)


func setCsvToBlank() {
	dbFile := "__db.csv"

	// Open file with truncate flag to clear contents
	fileHandle, err := os.OpenFile(dbFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		panic(err)
	}
	defer fileHandle.Close()

	// Nothing to write, file is now blank
}

func addToCsv(rowValues []string) int {
	dbFile := "__db.csv"

	// Open or create file in append mode
	fileHandle, err := os.OpenFile(dbFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer fileHandle.Close()

	// map-style replacement
	for i, value := range rowValues {
		rowValues[i] = `"` + strings.ReplaceAll(value, `"`, `\"`) + `"`
	}


	// Join row values into CSV format
	row := strings.Join(rowValues, ",") + "\n"

	// Write row
	if _, err := fileHandle.Write([]byte(row)); err != nil {
		panic(err)
	}

	// Read entire file to count lines
	rows, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}

	// Count number of newline characters = number of rows
	return bytes.Count(rows, []byte("\n"))
}

func trimRowValue(value string) string {
	return strings.Trim(
		strings.TrimSpace(value), 
		`"`,
	)
}


func trimRowValues(values []string) []string {
	unquoted := values;
	for i,value := range values {
		unquoted[i] = trimRowValue(value) // removes leading & trailing "
	}
	return unquoted;
}

func readCsvRow(rowId int) []string {
	dbFile := "__db.csv"
	// Read entire file to count lines
	rowBytes, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}
	rows := strings.Split(string(rowBytes), "\n")
	return trimRowValues(strings.Split(rows[rowId],","));
}

func getCsvRowCount() int {
	dbFile := "__db.csv"
	// Read entire file to count lines
	rowBytes, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}
	rows := strings.Split(string(rowBytes), "\n")
	return len(rows);
}

func readAllCsvRow() [][]string {

	numRows := getCsvRowCount()

	rows := make([][]string, numRows);
	for x := 0; x < numRows; x++ {
		rows[x] = readCsvRow(x);
	}
	return rows;
}

func getCsvColumnNumber(columnName string) int {
	columns := readCsvRow(0);

	for i,column := range columns {
		column = trimRowValue(column);
		if column == columnName {
			return i+1;
		}
	}
	panic("Column not found");
}

func updateCsvRow( rowId int, columnName string, value string) {
	rows := readAllCsvRow();

	columnNumber := getCsvColumnNumber(columnName);

	rows[rowId][columnNumber-1] = value;

	setCsvToBlank();
	// map-style replacement
	for _, row := range rows {
		addToCsv(row);
	}
}



func searchCsv(columnName string, value string) []int {
	rows := readAllCsvRow()

	columnNumber := getCsvColumnNumber(columnName)
	columnIndex := columnNumber - 1
	var rowIds []int

	for rowId := 1; rowId < len(rows); rowId++ {
		row := rows[rowId]
		if row[columnIndex] == value {
			rowIds = append(rowIds, rowId)
		}
	}
	return rowIds
}

func deleteCSVRow(deleteRowID int) {
    rows := readAllCsvRow()
    setCsvToBlank()

    for rowID, row := range rows {
        if rowID == deleteRowID {
            continue
        }
        addToCsv(row)
    }
}

func searchRoutine() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("Please input your query:")

	query1, _ := reader.ReadString('\n')
	query1 = trimRowValue(query1)

	rowIds1 := searchCsv("FirstName", query1)
	rowIds2 := searchCsv("LastName", query1)
	rowIds3 := searchCsv("Email", query1)

	// Combine all row IDs
	rowIds := append(append(rowIds1, rowIds2...), rowIds3...)

	fmt.Println(len(rowIds))

	for _, rowId := range rowIds {
		row := readCsvRow(rowId)
		fmt.Println(strings.Join(row, ","))
	}
}



func findFirstCsvRow(columnName string, value string) *int {
	rows := readAllCsvRow()

	columnNumber := getCsvColumnNumber(columnName)
	columnIndex := columnNumber - 1

	for rowId := 1; rowId < len(rows); rowId++ {
		row := rows[rowId]
		if row[columnIndex] == value {
			return &rowId
		}
	}
	return nil
}



func addNewRowRoutine() {

	fmt.Println(
		"Input email:",
	);

	reader := bufio.NewReader(os.Stdin)
	email1, _ := reader.ReadString('\n')
	email1 = strings.TrimSpace(email1)


	rowId := findFirstCsvRow("Email", email1);

	fmt.Println(
		"input firstname:",
	);


	name1, _ := reader.ReadString('\n')
	name1 = strings.TrimSpace(name1)
	
	fmt.Println(
		"input lastname:",
	);


	lastName1, _ := reader.ReadString('\n')
	lastName1 = strings.TrimSpace(lastName1)

	if rowId != nil {
		updateCsvRow(*rowId, "FirstName", name1)
		updateCsvRow(*rowId, "LastName", lastName1)
		return;
	}

	addToCsv([]string{email1, name1, lastName1});

	fmt.Println("Email/Name successfully added to csv")
	
}

// --- main() equivalent to the C# while loop section ---
func main() {
	rows := readAllCsvRow()

	if len(rows) == 0 {
		return
	}

	rowMaxLengths := make([]int, len(rows[0]))

	// Compute max cell width per column
	for _, row := range rows {
		for i, cell := range row {
			if len(cell) > rowMaxLengths[i] {
				rowMaxLengths[i] = len(cell)
			}
		}
	}

	totalLength := 0
	for _, l := range rowMaxLengths {
		totalLength += l
	}

	fmt.Printf(" %s\n", strings.Repeat("-", totalLength+3+len(rowMaxLengths)))

	for rowId, row := range rows {
		fmt.Print("| ")
		for i, cell := range row {
			padding := rowMaxLengths[i] - len(cell)
			fmt.Print(cell + strings.Repeat(" ", padding) + " |")
		}
		fmt.Print("\n")

		if rowId == 0 {
			fmt.Println(strings.Repeat("-", totalLength+5+len(rowMaxLengths)))
		}
	}

	fmt.Printf(" %s\n", strings.Repeat("-", totalLength+3+len(rowMaxLengths)))

	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print("Press 's' to search, 'a' to add or ESC to exit: ")
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)

		if input == "s" {
			searchRoutine()
			continue
		}

		if input == "a" {
			addNewRowRoutine()
			continue
		}

		if strings.ToLower(input) == "esc" {
			fmt.Println("Escape pressed. Exiting...")
			break
		}

		fmt.Println("Invalid command")
	}
}
